File: hash-table-easy/350.两个数组的交集-ii.py
#
# @lc app=leetcode.cn id=350 lang=python3
#
# [350] 两个数组的交集 II
#


from collections import defaultdict

# @lc code=start
class Solution:
    # 61/61 cases passed (56 ms)
    # Your runtime beats 75.58 % of python3 submissions
    # Your memory usage beats 25.61 % of python3 submissions (15 MB)
    def intersect(self, nums1: list[int], nums2: list[int]) -> list[int]:
        table = defaultdict(int)
        for i in nums1:
            table[i] += 1
        res = []
        for i in nums2:
            if i in table and table[i] != 0:
                table[i] -= 1
                res.append(i)
        return res


# @lc code=end

File: hash-table-easy/350.两个数组的交集-ii.rs
/*
 * @lc app=leetcode.cn id=350 lang=rust
 *
 * [350] 两个数组的交集 II
 */

// @lc code=start
impl Solution {
    // 61/61 cases passed (0 ms)
    // Your runtime beats 100 % of rust submissions
    // Your memory usage beats 38.46 % of rust submissions (2.1 MB)
    pub fn intersect(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        use std::collections::HashMap;
        let mut count_map = HashMap::new();
        for i in nums1.iter() {
            let count = count_map.entry(i).or_insert(0);
            *count += 1
        }
        let mut res = vec![];
        for i in nums2.iter() {
            if let Some(count) = count_map.get_mut(i) {
                if *count == 0 {
                    continue;
                }
                *count -= 1;
                res.push(*i);
            }
        }
        res
    }
}
// @lc code=end
